### PR TITLE
[Speculative Decoding] Support mtp super ultra overlap in pd-split mode with insert_task overlap

### PR DESCRIPTION
**Pull request by freeliuzc**
## Motivation

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191)

## Modifications
在PD分离模式下，特别是多DP，高并发情况下，两次推理间耗时较长，20+ms，且快慢卡严重（第一层deepep耗时异常）  
- v0\~v1(常规overlap+浅优化版本): 节省 MTP 和 Target 之间的空隙以及后处理空隙，约1~2ms，以及在插入 Decoder Query 时的 overlap行为
     - 问题：不能解决插入新 prefill 请求时的同步问题DP8 + 单 DP bsz 128，会持续插入 prefill 请求，且每个 DP 的插入会影响其他7个DP，导致加速被严重稀释between_step avg = 27ms 且快慢卡严重，deepep排队较长
- v2: insert_task_v1 在插入 prefill query 时完全异步处理，不占用GPU耗时between_step avg = 8ms 且快慢卡现象明显减轻 
- v3(完全体): 插入 prefill query 时，CPU 计算 max_batch_size当前只剩下save_output 一个必要同步between_step avg = 1.5ms 且几乎没有快慢卡现象相比原始版本，OTPS/解码性能均提升10%~15% 
- 总结：此模式在并发较大 / EP、DP较多情况下，性能加速明显理论集中式和 P节点都有加速预期老版本：

老版本：
bb2b98042e0bbdae7fe7d98a80c433a6  
新版本：
d82d065d18d5353831cf06e947628d4f

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [x] Provide accuracy results.
- [x] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -878,21 +882,29 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
...
+                    self._cached_real_bsz += 1
                     if self.speculative_decoding:
-                        # D speculate decode, seq_lens_this_time = length + 1
-                        self.share_inputs["seq_lens_this_time"][idx : idx + 1] = length + 1
-                        self.share_inputs["draft_tokens"][idx : idx + 1, 0 : length + 1] = paddle.to_tensor(
+                        # D speculate decode, [Target first token, MTP first draft token]
+                        async_set_value(self.share_inputs["seq_lens_this_time"][idx : idx + 1], 2)
+                        async_set_value(
+                            self.share_inputs["draft_tokens"][idx : idx + 1, 0:2],
                             request.draft_token_ids[0 : length + 1],
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py]: 已修复

File: fastdeploy/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -878,21 +882,29 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
                 if (
                     self.fd_config.scheduler_config.splitwise_role == "decode"
                 ):  # In PD, we continue to decode after P generate first token
-                    self.share_inputs["seq_lens_encoder"][idx : idx + 1] = 0
+                    # TODO: delete useless operation like this
+                    async_set_value(self.share_inputs["seq_lens_encoder"][idx : idx + 1], 0)
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py:886]: 已修复

File: fastdeploy/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -981,7 +1002,8 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
             self.sampler.apply_logits_processor(idx, logits_info, prefill_tokens)
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py:1010]: 已修复，这里在初始化 rename 了下

File: fastdeploy/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -818,9 +818,7 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
                 input_ids = prompt_token_ids + request.output_token_ids
                 prompt_len = len(prompt_token_ids)
                 # prompt_tokens
-                self.share_inputs["token_ids_all"][idx : idx + 1, :prompt_len] = np.array(
-                    prompt_token_ids, dtype="int64"
-                )
+                async_set_value(self.share_inputs["token_ids_all"][idx : idx + 1, :prompt_len], prompt_token_ids)
                 # generated_token_ids fill -1
                 self.share_inputs["token_ids_all"][idx : idx + 1, prompt_len:] = -1
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py:823]: 已修复

File: fastdeploy/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -981,7 +1002,8 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
             self.sampler.apply_logits_processor(idx, logits_info, prefill_tokens)
 
         self._process_mm_features(req_dicts)
-        if len(rope_3d_position_ids["position_ids_idx"]) > 0:
+
+        if len(rope_3d_position_ids["position_ids_idx"]) > 0 and self.enable_mm_runtime:
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py]: 已修复

File: fastdeploy/spec_decode/mtp.py [hunk 6]
```diff
@@ -483,28 +483,28 @@ def insert_tasks_v1(
...
+                async_set_value(
+                    self.model_inputs["block_tables"][idx : idx + 1, :encoder_block_num], request.block_tables
                 )
-                self.model_inputs["stop_flags"][idx : idx + 1] = False
-                self.model_inputs["batch_drop"][idx : idx + 1] = False
 
-                self.model_inputs["seq_lens_encoder"][idx : idx + 1] = length
+                async_set_value(self.model_inputs["stop_flags"][idx : idx + 1], False)
+                async_set_value(self.model_inputs["batch_drop"][idx : idx + 1], False)
+
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/spec_decode/mtp.py:507]: 已修复，在 async_set_value 里进行了判断

File: fastdeploy/spec_decode/mtp.py [hunk 7]
```diff
@@ -483,28 +483,28 @@ def insert_tasks_v1(
                 input_ids = request.prompt_token_ids + request.output_token_ids
 
                 self.model_inputs["input_ids_len"][idx] = length - 1
-                self.model_inputs["pre_ids"][idx : idx + 1] = -1
+                async_set_value(self.model_inputs["pre_ids"][idx : idx + 1], -1)
                 self.model_inputs["input_ids"][idx : idx + 1, : length - 1] = self.target_model_inputs["input_ids"][
                     idx : idx + 1, 1:length
                 ]
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/spec_decode/mtp.py:492]: 已修复，仅有在 hybrid_mode 里赋值，后续会改成 GPU 变量

File: fastdeploy/eplb/async_expert_loader.py [hunk 8]
```diff
@@ -24,8 +24,20 @@
...
+        import importlib.metadata as _meta
+
+        _cuda_ver = _meta.version("cuda-bindings")
+    _cuda_major = int(_cuda_ver.split(".")[0])
+    if _cuda_major >= 13:
+        from cuda.bindings import runtime as cudart
+    else:
+        from cuda import cudart
+except (ImportError, Exception):
     cudart = None
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/eplb/async_expert_loader.py:44]: 这里不是强需求，添加 warning

File: fastdeploy/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -878,21 +882,37 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
...
+                        if len(draft_tokens_to_write) != 2:
+                            raise ValueError(
+                                "Expected at least 2 draft tokens for speculative suffix decode, "
+                                f"but got {len(draft_tokens_to_write)} for request {request.request_id}."
+                            )
+                        async_set_value(
+                            self.share_inputs["draft_tokens"][idx : idx + 1, 0:2],
+                            draft_tokens_to_write,
                         )
+                        async_set_value(self.share_inputs["seq_lens_this_time"][idx : idx + 1], 2)
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py]: 确实，之前以一种奇怪的方式在正确运行

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 10]
```diff
@@ -116,36 +116,33 @@
...
+    elif isinstance(src, (list, np.ndarray)):
+        dtype_str = str(tgt.dtype).split(".")[1]
+        if isinstance(src, list):
+            src = np.array(src, dtype=dtype_str if dtype_str != "bfloat16" else "float32")
+        if current_platform.is_cuda():
             if str(src.dtype) != dtype_str:
                 srt_tensor = paddle.empty(tgt.shape, dtype=str(src.dtype))
                 src = custom_numpy_to_tensor(src, srt_tensor)
             else:
                 return custom_numpy_to_tensor(src, tgt)
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/model_executor/pre_and_post_process.py:132]: 这里后续有其他PR继续完善

File: fastdeploy/spec_decode/mtp.py [hunk 11]
```diff
@@ -483,28 +483,32 @@ def insert_tasks_v1(
...
-                self.model_inputs["input_ids_cpu"][idx : idx + 1, : length - 1] = self.target_model_inputs[
-                    "input_ids"
-                ][idx : idx + 1, 1:length].cpu()
+                # TODO: use token_all_ids replace with input_ids_cpu
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/spec_decode/mtp.py:493]: 后续会统一到 GPU tensor

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 12]
```diff
@@ -116,36 +116,33 @@
...
-
-    def async_set_value(tgt, src):
-        if isinstance(src, (int, float, bool)):
-            src = paddle.full(tgt.shape, fill_value=src, dtype=tgt.dtype)
-        elif isinstance(src, (list, np.array)):
-            dtype_str = str(tgt.dtype).split(".")[1]
-            if isinstance(src, list):
-                src = np.array(src, dtype=dtype_str if dtype_str != "bfloat16" else "float32")
+
+def async_set_value(tgt, src):
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/model_executor/pre_and_post_process.py:120]: 这里暂时先兼容下其他平台

File: fastdeploy/worker/gpu_model_runner.py [hunk 13]
```diff
@@ -878,21 +882,37 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
                 if (
                     self.fd_config.scheduler_config.splitwise_role == "decode"
                 ):  # In PD, we continue to decode after P generate first token
-                    self.share_inputs["seq_lens_encoder"][idx : idx + 1] = 0
+                    # TODO: delete useless operation like this
+                    async_set_value(self.share_inputs["seq_lens_encoder"][idx : idx + 1], 0)
                     self.exist_prefill_flag = False
-                    self._cached_launch_token_num = -1
+                    if self._cached_launch_token_num != -1:
+                        token_num_one_step = (
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py:889]: 已添加单测

File: fastdeploy/model_executor/xpu_pre_and_post_process.py [hunk 14]
```diff
@@ -55,6 +55,28 @@
...
 
+def async_set_value(tgt, src):
+    if isinstance(src, (int, float, bool)):
+        src = paddle.full(tgt.shape, fill_value=src, dtype=tgt.dtype)
+    elif isinstance(src, (list, np.ndarray)):
+        dtype_str = str(tgt.dtype).split(".")[1]
+        if isinstance(src, list):
+            src = np.array(src, dtype=dtype_str if dtype_str != "bfloat16" else "float32")
+            # TODO: support async_numpy_to_tensor
+            src = paddle.to_tensor(src, dtype=tgt.dtype)
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/model_executor/xpu_pre_and_post_process.py]: 已修复

File: fastdeploy/spec_decode/mtp.py [hunk 15]
```diff
@@ -483,28 +486,32 @@ def insert_tasks_v1(
                 input_ids = request.prompt_token_ids + request.output_token_ids
 
                 self.model_inputs["input_ids_len"][idx] = length - 1
-                self.model_inputs["pre_ids"][idx : idx + 1] = -1
+                async_set_value(self.model_inputs["pre_ids"][idx : idx + 1], -1)
                 self.model_inputs["input_ids"][idx : idx + 1, : length - 1] = self.target_model_inputs["input_ids"][
```
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/spec_decode/mtp.py:490]: 忽略，这里可以异步

File: fastdeploy/worker/gpu_model_runner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (freeliuzc) [reply fastdeploy/worker/gpu_model_runner.py:823]: 忽略，这里可以异步

**Review by Sunny-bot1 — APPROVED** [review]

**Review by yuanlehome — APPROVED** [review]

**Review by ming1753 — APPROVED** [review]
LGTM

**Review by heavengate — APPROVED** [review]
